### PR TITLE
Feature/deploy with flags

### PR DESCRIPTION
**Pull request by AyushyaChitransh**
The branch is ready to be merged.

### CODE REVIEW

**Comment by SolomonShorser-OICR** [comment]
When I run with `-d all`, I get this error at the end of the script:
```
#######################################################################
# Filename:   gk_current.sql.gz
#######################################################################
rm: cannot remove 'mysql/tomcat_data/gk_current.sql.gz': Permission denied
```

When I run with `-u all` I get this error at the end of the script:
```
./deploy.sh: line 375: i++: command not found
```

**Comment by SolomonShorser-OICR** [comment]
When I run with `-b`, it fails to build ReactomeRESTfulAPI with this message:

`The specified file '/gitroot/CuratorTool/reactome.jar' not exists`

**Comment by SolomonShorser-OICR** [comment]
When I run `./deploy.sh -b select` and build a single application, it builds the application but then it tries to run the containers. I was expecting `-b` to build but not to run. This behaviour should be documented as `Build essential java web applications, and run the containers` or something like that. Or, change `-b` to not run the containers.

**Comment by AyushyaChitransh** [comment]
By default, only some applications will be built, and rest of them will not be built, and this is the message being shown for the applications which are not getting built.

> Unknown application state <Application_name>

The application states are getting stored in the local shell environment. Earlier I was thinking of writing those changes in files, but that would be more permanent kind of thins, so now the application states are temporarily saved in shell. From there, build_webapps.sh takes those states and maps them to the variables defined in `build_webapps.env` file. This finally tells `maven_builds.sh` which applications are to be built.

Implemented these changes in commit f18b488

**Comment by AyushyaChitransh** [comment]
I think we should make download, update and build flags exclusive. They should fulfill only _their_ purpose, and not do anything else. This seems more intuitive.

**Comment by SolomonShorser-OICR** [comment]
I agree.

**Comment by AyushyaChitransh** [comment]
Also, we would be including in the documents' instructions, that for first use, users are required to run `-b all` so that all the required repositories and files are built. Like RESTfulAPI depends on CuratorTools. So CuratorTools needs to be built first

**Comment by SolomonShorser-OICR** [comment]
That's a good point. I didn't know I need to run `-b all` the first time.

**Comment by SolomonShorser-OICR** [comment]
I also notice that every time a build fails, I now get this message the next time I run the deployment script:
```
Running Analysis Service requires a working database
Error response from daemon: network with name isolated_nw already exists
docker: Error response from daemon: Conflict. The container name "/mysql-for-webapps" is already in use by container...
```
I then need to manually remove the container.
This happens even if I do `-b select` and I choose ContentService. I don't think this used to happen to me.

**Comment by SolomonShorser-OICR** [comment]
This commit https://github.com/reactome/container/commit/5d501454dcab1aa803bea6b98cd48b6692bb0a65 should resolve the issues with SBMLExporter. The project was updated recently, so if there is new code in the java-application-build container, the POM needs to updated as well.

**Comment by AyushyaChitransh** [comment]
Resolved the issue of conflicting containers and running only deploy.sh would start container while other options would not trigger the container.

Additionally, users would be able run the containers by supplying  `run` argument. But now I see that, `run` should be implemented as a flag, not an argument.

**Comment by AyushyaChitransh** [comment]
Resolved the issue of conflicting containers and running only deploy.sh would start container while other options would not trigger the startup of container.

Additionally, users would be able run the containers by supplying  `-run` flag.

**Comment by SolomonShorser-OICR** [comment]
Here's something unusual: If I delete a file (such as mysql/tomcat_data/gk_current.sql.gz) then I can't get the script to download it again. If I do `./deploy.sh -d` I get this error:

```
#######################################################################
# Filename:   gk_current.sql.gz
#######################################################################
Old file deleted, downloading new one...
mysql/tomcat_data/gk_current.sql.gz: Permission denied
```
If I use the `-u` flag, I get this:
```
File mysql/tomcat_data/gk_current.sql.gz does not exist. Will download now.
mysql/tomcat_data/gk_current.sql.gz: Permission denied
```
I tried deleting another file (the solr index - solr/data/solr_data.tgz) and the script downloaded it fine with `-d` or `-u`.

File: deploy.sh [hunk 1]
```diff
@@ -69,98 +116,339 @@ if [ $? -eq 0 ]; then
...
-  cd ./java-application-builder
-  bash ./build_webapps.sh |& tee ../logs/build_webapps.log
-  cd ..
-  echo "Reactome-app-builder exits here."
+    -h | --help )
+      # Displaying help
+      echo "$usage"
+      exit 0
+      ;;
+    -run )
```
Comment (SolomonShorser-OICR) [inline deploy.sh]: Could this be changed to `-r | --run`? Also needs to be documented in the help text.

**Comment by SolomonShorser-OICR** [comment]
I'm going to try to resolve the merge conflicts. Wish me luck!

**Comment by SolomonShorser-OICR** [comment]
I can't find any other outstanding issues with the branch, I'll merge back into develop. If there's anything else, it can be fixed on another branch.

**Pull request merged**